crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,808 @@
+            let block1 = random_block(1, Some(genesis.hash), None, Some(0));
+
+            insert_blocks(env.db.as_ref(), [&genesis, &block1].into_iter());
+
+            let _ = spawn_consensus_engine(consensus_engine);
+
+            let rx = env.send_forkchoice_updated(ForkchoiceState {
+                head_block_hash: H256::random(),
+                finalized_block_hash: block1.hash,
+                ..Default::default()
+            });
+            let expected_result = ForkchoiceUpdated::from_status(PayloadStatusEnum::Syncing);
+            assert_matches!(rx.await, Ok(Ok(result)) => assert_eq!(result, expected_result));
+
+            let rx = env.send_forkchoice_updated(ForkchoiceState {
+                head_block_hash: block1.hash,
+                finalized_block_hash: block1.hash,
+                ..Default::default()
+            });
+            let expected_result = ForkchoiceUpdated::from_status(PayloadStatusEnum::Invalid {
+                validation_error: ExecutorError::BlockPreMerge.to_string(),
+            })
+            .with_latest_valid_hash(H256::zero());
+            assert_matches!(rx.await, Ok(Ok(result)) => assert_eq!(result, expected_result));
+        }
+    }
+
+    mod new_payload {
+        use super::*;
+        use reth_interfaces::{
+            executor::Error as ExecutorError, test_utils::generators::random_block,
+        };
+        use reth_primitives::{Hardfork, U256};
+        use reth_provider::test_utils::blocks::BlockChainTestData;
+
+        #[tokio::test]
+        async fn payload_known() {
+            let chain_spec = Arc::new(
+                ChainSpecBuilder::default()
+                    .chain(MAINNET.chain)
+                    .genesis(MAINNET.genesis.clone())
+                    .paris_activated()
+                    .build(),
+            );
+            let (consensus_engine, env) = setup_consensus_engine(
+                chain_spec,
+                VecDeque::from([Ok(ExecOutput { done: true, stage_progress: 0 })]),
+                Vec::default(),
+            );
+
+            let genesis = random_block(0, None, None, Some(0));
+            let block1 = random_block(1, Some(genesis.hash), None, Some(0));
+            let block2 = random_block(2, Some(block1.hash), None, Some(0));
+            insert_blocks(env.db.as_ref(), [&genesis, &block1, &block2].into_iter());
+
+            let mut engine_rx = spawn_consensus_engine(consensus_engine);
+
+            // Send forkchoice
+            let rx = env.send_forkchoice_updated(ForkchoiceState {
+                head_block_hash: block1.hash,
+                finalized_block_hash: block1.hash,
+                ..Default::default()
+            });
+            let expected_result =
+                ForkchoiceUpdated::new(PayloadStatus::from_status(PayloadStatusEnum::Syncing));
+            assert_matches!(rx.await, Ok(Ok(result)) => assert_eq!(result, expected_result));
+
+            // Send new payload
```

Comment:
do we not want a scenario where new payload is first, and then fork choice state?

Reply:
that's a good one. added `new_payload_before_forkchoice`